### PR TITLE
Transforms documentation clean-up

### PR DESCRIPTION
**Pull request by voldemortX**
A documentation clean-up that concerns` transform.py functional.py functional_pil.py functional_tensor.py` for issue #3071.
@datumbox After some initial checking, there are somethings I'm not certain:

1. Does the `Sequence` type represent only list and tuple in torchvision? I find that in Python other types (e.g. strings) can also count as a Sequence. Maybe we should clear it up in the docs, like maybe just say list or tuple.
2. Does the Posterize operation also only support uint8 images in PIL? I'm inclined to add "only support L or RGB" to all related transforms like I did with invert(). But maybe we can just let PIL throw the OSError.
3. Should we also add "This transform supports both PIL and Tensor, ..." in the documentation for AutoAugment as well?
4. Maybe we could delete the private function docs, as mentioned by @vfdev-5 [here](https://github.com/pytorch/vision/pull/3069#pullrequestreview-541808920).

And there are some other things I found after cleaning the docs up, maybe they can be addressed in the future:

1. In Pad (or RandomCrop, it also uses padding), it seems tensors only support single value fill. Also I'm not sure whether str is really needed here by PIL, what is the use case here?
2. Some type checking is also Sequence, maybe we should check explicitly for tuple and list. Or is this the intended implementation?

### CODE REVIEW

File: torchvision/transforms/functional.py [hunk 1]
```diff
@@ -1180,8 +1181,8 @@ def invert(img: Tensor) -> Tensor:
     Args:
         img (PIL Image or Tensor): Image to have its colors inverted.
             If img is a Tensor, it is expected to be in [..., H, W] format,
-            where ... means it can have an arbitrary number of trailing
-            dimensions.
+            where ... means it can have an arbitrary number of leading dimensions.
+            For PIL images, only mode "L" and "RGB" is supported.
```
Comment (vfdev-5) [inline torchvision/transforms/functional.py:1185]: How about "F" mode ?

**Review by vfdev-5 — COMMENTED** [review]
Thanks for the PR @voldemortX 

> Does the Sequence type represent only list and tuple in torchvision? I find that in Python other types (e.g. strings) can also count as a Sequence. Maybe we should clear it up in the docs, like maybe just say list or tuple.

Yes, sequence is also a string. I'd say here it is implicitly assumed that string input for, for example, size="abc" is completely wrong... also `isinstance(size, Sequence)` is a bit shorter then `isinstance(size, (list, tuple))`.


> Maybe we could delete the private function docs

I'd say yes

**Comment by datumbox** [comment]
To add on @vfdev-5:

> 2. Does the Posterize operation also only support uint8 images in PIL? I'm inclined to add "only support L or RGB" to all related transforms like I did with invert(). But maybe we can just let PIL throw the OSError.

Yes it's only supported in PIL when the image is uint8. In this case it's important to clarify to the users that the tensor should be of data type uint8. Currently an image can be RGB but represented as floats. If that's the case, the transform won't work.

> 3. Should we also add "This transform supports both PIL and Tensor, ..." in the documentation for AutoAugment as well?

Not necessary because supporting a transform in both backends is our default. We only specific when something is not supported in any of the two backends.

File: torchvision/transforms/functional.py (hunk 1 above)
Comment (voldemortX) [reply torchvision/transforms/functional.py:1185]: It seems these ops all use the [PIL.ImageOps._lut()](https://pillow.readthedocs.io/en/stable/_modules/PIL/ImageOps.html?highlight=_lut#), and it only takes L or RGB at the moment.

Comment (datumbox) [reply torchvision/transforms/functional.py:1185]: I'm not sure we currently have any tests for the specific mode or that there was intention to support them.

**Comment by voldemortX** [comment]
> Not necessary because supporting a transform in both backends is our default. We only specific when something is not supported in any of the two backends.

Maybe we should remove that support both backend statement for other docs as well?

File: torchvision/transforms/functional.py [hunk 2]
```diff
@@ -1117,7 +1117,8 @@ def erase(img: Tensor, i: int, j: int, h: int, w: int, v: Tensor, inplace: bool
 def gaussian_blur(img: Tensor, kernel_size: List[int], sigma: Optional[List[float]] = None) -> Tensor:
     """Performs Gaussian blurring on the img by given kernel.
     The image can be a PIL Image or a Tensor, in which case it is expected
-    to have [..., H, W] shape, where ... means an arbitrary number of leading dimensions
+    to have [..., H, W] shape, where ... means an arbitrary number of leading dimensions.
+    If input is a PIL image, it will be first converted to Tensor for this operation, then converted back.
```
Comment (datumbox) [inline torchvision/transforms/functional.py:1121]:
```suggestion
```
The conversion from PIL => Tensor => PIL is an implementation detail which I don't think should be part of the documentation.

**Comment by voldemortX** [comment]
> Yes, sequence is also a string. I'd say here it is implicitly assumed that string input for, for example, size="abc" is completely wrong... also `isinstance(size, Sequence)` is a bit shorter then `isinstance(size, (list, tuple))`.

Does that need to be like unified in the doc to use sequence everywhere or int and tuple everywhere? @vfdev-5

File: torchvision/transforms/functional_tensor.py [hunk 3]
```diff
@@ -678,7 +678,7 @@ def pad(img: Tensor, padding: List[int], fill: int = 0, padding_mode: str = "con
             this is the padding for the left, top, right and bottom borders
             respectively. In torchscript mode padding as single int is not supported, use a tuple or
             list of length 1: ``[padding, ]``.
-        fill (int): Pixel fill value for constant fill. Default is 0.
+        fill (int, float): Pixel fill value for constant fill. Default is 0.
```
Comment (datumbox) [inline torchvision/transforms/functional_tensor.py:681]: As discussed in comments, let's go ahead and remove all docs from `functional_tensor.py` and `torchvision/transforms/functional_pil.py`

File: torchvision/transforms/transforms.py [hunk 4]
```diff
@@ -1040,6 +1044,9 @@ def __repr__(self):
 
 class ColorJitter(torch.nn.Module):
     """Randomly change the brightness, contrast, saturation and hue of an image.
+    The image can be a PIL Image or a Tensor, in which case it is expected
+    to have [..., H, W] shape, where ... means an arbitrary number of leading
+    dimensions. This transform does not support torchscript.
```
Comment (datumbox) [inline torchvision/transforms/transforms.py:1049]:
```suggestion
    dimensions.
```

I think it's supported. See the tests.

**Comment by vfdev-5** [comment]
> > Yes, sequence is also a string. I'd say here it is implicitly assumed that string input for, for example, size="abc" is completely wrong... also `isinstance(size, Sequence)` is a bit shorter then `isinstance(size, (list, tuple))`.
> 
> Does that need to be like unified in the doc to use sequence everywhere or int and tuple everywhere? @vfdev-5

Yes, in the docs it is nice to unify the type description. In the code this may be not possible due to torchscript...

File: torchvision/transforms/transforms.py [hunk 5]
```diff
@@ -1483,7 +1488,7 @@ def __repr__(self):
 
 
 class RandomErasing(torch.nn.Module):
-    """ Randomly selects a rectangle region in an image and erases its pixels.
+    """ Randomly selects a rectangle region in an torch Tensor image and erases its pixels.
```
Comment (datumbox) [inline torchvision/transforms/transforms.py:1491]: Good catch. Perhaps be even more specific and mention that it does not support PIL?

File: torchvision/transforms/transforms.py [hunk 6]
```diff
@@ -1607,7 +1612,8 @@ class GaussianBlur(torch.nn.Module):
     """Blurs image with randomly chosen Gaussian blur.
     The image can be a PIL Image or a Tensor, in which case it is expected
     to have [..., C, H, W] shape, where ... means an arbitrary number of leading
-    dimensions
+    dimensions.
+    If input is a PIL image, it will be first converted to Tensor for this operation, then converted back.
```
Comment (datumbox) [inline torchvision/transforms/transforms.py:1616]:
```suggestion
```

Same as above, implementation detail.

File: torchvision/transforms/transforms.py [hunk 7]
```diff
@@ -1707,6 +1713,7 @@ class RandomInvert(torch.nn.Module):
     The image can be a PIL Image or a torch Tensor, in which case it is expected
     to have [..., H, W] shape, where ... means an arbitrary number of leading
     dimensions.
+    For PIL images, only mode "L" and "RGB" is supported.
```
Comment (datumbox) [inline torchvision/transforms/transforms.py:1716]: I think that's true for both PIL and Tensor.

**Review by datumbox — COMMENTED** [review]
Thanks a lot for the PR. These are much needed changes that hopefully will make the documentation clearer. :)

I left a couple of comments. Let me know what you think.

**Comment by datumbox** [comment]
> Yes, in the docs it is nice to unify the type description. In the code this may be not possible due to torchscript..

I agree. Torchscript might give you headaches. Perhaps ensure that our doc is consistent everywhere using Sequences in this PR and follow up with the proposed changes on another one if TorchScript plays nice?

File: torchvision/transforms/functional.py (hunk 2 above)
Comment (voldemortX) [reply torchvision/transforms/functional.py:1121]: Yes, you are right. I initially thought maybe some people who are concerned about speed might want to know about it.

File: torchvision/transforms/transforms.py (hunk 4 above)
Comment (voldemortX) [reply torchvision/transforms/transforms.py:1049]: You are right! I put that there the moment I saw a jit.unused... in _check_input() I thought that means unsupported. Since private methods can support jit like this, I think some of my previous code might needs a rewrite.

File: torchvision/transforms/transforms.py (hunk 7 above)
Comment (voldemortX) [reply torchvision/transforms/transforms.py:1716]: Let me check if my understanding is correct: it should be that all methods that use _lut() supports only L and RGB for PIL, uint8 for tensor?

**Comment by voldemortX** [comment]
I made some further changes based on above reviews. Let me know if I forgot something or got something wrong!
@vfdev-5 @datumbox

Also TODOs in future PRs maybe:
1. Try check only for number and sequence in the code.
2. Try unify tensor and PIL Image in pad().

**Review by datumbox — CHANGES_REQUESTED** [review]
@voldemortX Apologies if I was not very clear. It's not our intention to mark the specific files with underscores at this point. 

A very bad side-effect of doing so is we lose the entire git-blame history, so such changes need to be justified. Please revert the name change and I'll check again your PR once it's done.

**Comment by voldemortX** [comment]
> @voldemortX Apologies if I was not very clear. It's not our intention to mark the specific files with underscores at this point.
> 
> A very bad side-effect of doing so is we lose the entire git-blame history, so such changes need to be justified. Please revert the name change and I'll check again your PR once it's done.

Sorry... Should I revert all commits?

File: torchvision/transforms/transforms.py (hunk 7 above)
Comment (datumbox) [reply torchvision/transforms/transforms.py]: Both require that they are L or RGB. See the entering invert() functional method:

https://github.com/pytorch/vision/blob/90645ccd0e774ad76200245e32222a23d09f2312/torchvision/transforms/functional.py#L1177-L1178

**Comment by datumbox** [comment]
> Sorry... Should I revert all commits after 08fba9b?

No worries mate. :) You can just undo the filename change in a new commit or revert the specific commit or however else you prefer. When we merge the PRs, the history is squashed so there will be no side-effects on master.

**Comment by voldemortX** [comment]
@datumbox Is this new commit ok?

File: torchvision/transforms/transforms.py (hunk 7 above)
Comment (voldemortX) [reply torchvision/transforms/transforms.py]: Should I specify the related docs (invert, solarize, etc.) like, for example:

"""Invert the colors of an RGB/grayscale image.

    Args:
        img (PIL Image or Tensor): L or RGB image to have its colors inverted.
            If img is a Tensor, it is expected to be in [..., 1 or 3, H, W] format,
            where ... means it can have an arbitrary number of leading dimensions.

    Returns:
        PIL Image or Tensor: Color inverted image.
"""

**Review by datumbox — APPROVED** [review]
@voldemortX Awesome work! Thanks a lot for spending the time not just removing the duplicate docs, but finding out what is currently supported and what's not. This is a major improvement on our documentation.

**Comment by voldemortX** [comment]
After seeing this issue #3206 , we might need to also clarify the doc for erase(). I think it does support batch input (although the same randomness across the whole batch) right?

**Comment by datumbox** [comment]
Yes it does but the input needs to be a tensor, not a PIL image.
